### PR TITLE
[WIP] Key manager permissions part-1

### LINKED ISSUE #5753: Granted keys should not be transferable/cancellable by default
In the past we have discussed challenges around granted keys. I believe that, by default, keys which have been created thru the `grantKeys` step should not be transferable/cancellable by their owner.

This should leverage the [work done on this issue](https://github.com/unlock-protocol/unlock/issues/5557).

However the grantor (or lock owner) should be able to revoke that limitation, if they want.

### PR DESCRIPTION
**Pull request by nfurfaro**
# Description
This is part 1/3 for adding new keyManager permissions. Splitting up to simplify review.
- This PR supercedes #5965
- The next PR will include an update to setApprovalForAll and address the permissions around cancelation/refunds.

# Issues

Fixes #5753
Refs #

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [x] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: smart-contracts/contracts/mixins/MixinGrantKeys.sol [hunk 1]
```diff
@@ -35,6 +37,12 @@ contract MixinGrantKeys is
 
       _assignNewTokenId(toKey);
       _recordOwner(recipient, toKey.tokenId);
+      // Only assign the KeyManager for new keys
+      if(keyManagerOf[toKey.tokenId] == address(0)) {
```
Comment (nfurfaro) [inline smart-contracts/contracts/mixins/MixinGrantKeys.sol]: hmm. I think this needs work.

Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinGrantKeys.sol]: The intent was to prevent the use of `grantKeys` to be able to "take-over" an existing user's key. EG:
you have a valid key. I grant you a new key (which extends yours) but I also set myself as the keyManager for your key.

Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinGrantKeys.sol]: but there's already a check for this above, so I will push a small change to simplify this.

File: smart-contracts/contracts/mixins/MixinKeys.sol [hunk 2]
```diff
@@ -49,6 +51,11 @@ contract MixinKeys is
   // Addresses are never removed by design to avoid abuses around referals
   address[] public owners;
 
+  // A given key has both an owner and a manager.
+  // The owner and manager may be the same address
+  // Each key can have at most 1 keyManager, but may have no manager.
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinKeys.sol]: What does no manager mean? Does that imply the keyOwner is the manager at that time? I think that could make sense but the current implementation I think effectively disables the key if the manager is 0

Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: Also... not sure on this. But should keyManager be a field of the Key struct?

File: smart-contracts/contracts/mixins/MixinKeys.sol [hunk 3]
```diff
@@ -228,6 +243,35 @@ contract MixinKeys is
     return _ownerOf[_tokenId];
   }
 
+  /**
+  * @notice Get the key manager for the given tokenId
+  * @return The address of the manager for the given key
+  */
+  function getKeyManagerOf(
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinKeys.sol]: In general do we prefer getters like this or public variables?  We should aim to pick one for consistency.

File: smart-contracts/contracts/mixins/MixinTransfer.sol [hunk 4]
```diff
@@ -48,10 +48,12 @@ contract MixinTransfer is
   function shareKey(
     address _to,
     uint _tokenId,
-    uint _timeShared
+    uint _timeShared,
+    address _keyManager
   ) public
     onlyIfAlive
     onlyKeyOwnerOrApproved(_tokenId)
+    onlyKeyManager(_tokenId)
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinTransfer.sol]: I think having multiple modifiers using the same params is a little hard to follow.  Can we merge these into a single modifier?

File: smart-contracts/contracts/mixins/MixinKeys.sol (hunk 2 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: I don't feel strongly about this either way, but I thought we had discussed and decided that address(0) was a valid keyManager address.
The key would still work, but not be transferrable/shareable.

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 4 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: This change is made already, just not part of this PR. (It's in with the changes to the `approve` & `setApprovalForAll` permissions.

File: smart-contracts/contracts/mixins/MixinKeys.sol (hunk 2 above)
Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: Okay that's a fair use case. Just ensure there's a comment to that effect - maybe on the setter. Something like 'setting to 0 disables transfers for this key'.

Hmm begs a question. What if I buy with CC, Unlock is the key manager. They set the keyManager to 0. A few months later I buy again - this time with crypto. ..but I still cannot transfer? I'll never get those permissions back?

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 4 above)
Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: okay that's fine. maybe I should focus on that one first. i'm not sure having both like this is valid

File: smart-contracts/contracts/mixins/MixinKeys.sol (hunk 2 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: good point.of course, Unlock should probably have set themselves as the keyManager, but maybe supporting the zero address as a keyManager is overly complicated for the limited value it brings.

Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: Maybe on purchase, if the user does not have a key or the key is expired ATM - they become the keyManager.  e.g. if the first purchase was with CC and then expired before we updated the keyManager - it seems that on next purchase the buyer reclaim that role.

File: smart-contracts/contracts/mixins/MixinKeys.sol (hunk 3 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: yup. I think I should make this a public variable.

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 4 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: NOw that the way forward with approvals is clear I could add it to this pr... it's dependent on this one anyway and doesn't change much. It should simplify review.

File: smart-contracts/contracts/mixins/MixinKeys.sol (hunk 2 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: Yeah that makes sense. Also, just saw this:
>Also... not sure on this. But should keyManager be a field of the Key struct?

File: smart-contracts/contracts/mixins/MixinKeys.sol (hunk 3 above)
Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: I lean that way too. But a benefit of using getters instead is we can make variables private and therefore any use by other mixins is known to be contained to functions inside that contract.

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 4 above)
Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: Okay that works. I just hesitate to merge to master with something like this. An alternate approach we could consider is PRing small parts into a `sm-dev` branch. Then once the entire feature is working we PR that branch to master for a final sanity check.

File: smart-contracts/contracts/mixins/MixinKeys.sol (hunk 2 above)
Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: Yea I don't really know the tradeoffs. Begs the question - should we have a Key struct at all? And does the answer change knowing Solidity 0.6 is coming which includes abiV2 (which means we can publicly expose structs)

Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: Can we make that a new issue to track it? we can always refactor if we decide to change things up.

Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: >Maybe on purchase, if the user does not have a key or the key is expired ATM - they become the keyManager. e.g. if the first purchase was with CC and then expired before we updated the keyManager - it seems that on next purchase the buyer reclaim that role.

this has now been fixed.

File: smart-contracts/contracts/mixins/MixinGrantKeys.sol [hunk 5]
```diff
@@ -35,6 +37,12 @@ contract MixinGrantKeys is
 
       _assignNewTokenId(toKey);
       _recordOwner(recipient, toKey.tokenId);
+
+      // Only assign the KeyManager for new or expired keys
+      if(toKey.expirationTimestamp <= block.timestamp) {
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinGrantKeys.sol]: Hmm are you sure?  It seems this puts the CC use case at risk.

File: smart-contracts/contracts/mixins/MixinKeys.sol [hunk 6]
```diff
@@ -228,6 +243,24 @@ contract MixinKeys is
...
+  * setting to address(0) disables transfers for this key until it expires
+  * */
+  //
+  function setKeyManagerOf(
+    uint _tokenId,
+    address _keyManager
+  ) public
+    isKey(_tokenId)
+  {
+    require((msg.sender == keyManagerOf[_tokenId]) || (isLockManager(msg.sender) == true), 'SETKEYMANAGEROF_ACCESS_DENIED');
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinKeys.sol]: For consistent syntax / style don't include `== true` when checking a bool. The params here are also not needed.

File: smart-contracts/contracts/mixins/MixinKeys.sol (hunk 2 above)
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinKeys.sol]: Could you add a comment to clarify what no manager means?  I thought we were talking about removing that. But the code still supports no manager.

I think I prefer no manager means the keyOwner.  Then we only set it if it's different - e.g. it's 0 for a normal purchase.

File: smart-contracts/contracts/mixins/MixinGrantKeys.sol [hunk 7]
```diff
@@ -35,6 +37,12 @@ contract MixinGrantKeys is
...
+        keyManagerOf[toKey.tokenId] = keyManager;
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinGrantKeys.sol]: If we want to say keyManager is not 0, we should add a require

File: smart-contracts/contracts/mixins/MixinTransfer.sol [hunk 8]
```diff
@@ -85,6 +86,9 @@ contract MixinTransfer is
       _assignNewTokenId(toKey);
       iDTo = toKey.tokenId;
       _recordOwner(_to, iDTo);
+      // Assign the KeyManager
+      keyManagerOf[toKey.tokenId] = _keyManager;
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinTransfer.sol]: This is not consistent with purchase. What if the iDTo owns an expired key? We should be consistent or simplify

File: smart-contracts/contracts/mixins/MixinTransfer.sol [hunk 9]
```diff
@@ -48,10 +48,11 @@ contract MixinTransfer is
   function shareKey(
     address _to,
     uint _tokenId,
-    uint _timeShared
+    uint _timeShared,
+    address _keyManager
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinTransfer.sol]: Do we really want this? I think we can simplify and exclude keyManager here.

Thin keyManager does not change in this scenario. And no keyManager means that the keyOwner has those rights.

File: smart-contracts/contracts/mixins/MixinTransfer.sol [hunk 10]
```diff
@@ -128,6 +132,8 @@ contract MixinTransfer is
     if (toKey.tokenId == 0) {
       toKey.tokenId = fromKey.tokenId;
       _recordOwner(_recipient, toKey.tokenId);
+      keyManagerOf[toKey.tokenId] = _recipient;
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinTransfer.sol]: Again what about the expired scenario?

File: smart-contracts/contracts/mixins/MixinGrantKeys.sol (hunk 5 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinGrantKeys.sol]: What's the risk you see here? 
What I want to avoid is this scenario:
- you have a valid key which you purchased, and for which you are the keyManager
- I (the keyGranter) come along and grant you a new key which extends your existing key, but I also set myself as the keyManager of your key, in effect stealing it from you.

File: smart-contracts/contracts/mixins/MixinKeys.sol (hunk 6 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: I will remove the `== true`. thx for the consistency reminder. 
not sure about the params though... `isLockManager` takes an address and `keyManagerOf` needs a uint. I know I must be missing something very obvious.

File: smart-contracts/contracts/mixins/MixinGrantKeys.sol (hunk 5 above)
Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinGrantKeys.sol]: I buy with CC. I see a pending tx from Unlock protocol. I frontrun a crypto purchase. Then unlock buys and I cancel and refund, and chargeback - potentially doubling my investment

File: smart-contracts/contracts/mixins/MixinKeys.sol (hunk 2 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: Looking back over our discussion, we did discuss the merits of both keeping/removing this.
>I think I prefer no manager means the keyOwner. Then we only set it if it's different - e.g. it's 0 for a normal purchase.

 I support this and will modify so that if 0 is passed as the keyManager address, then the  keyOwner is also the manager.

File: smart-contracts/contracts/mixins/MixinGrantKeys.sol (hunk 5 above)
Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinGrantKeys.sol]: No that's not exactly right.  I think we should just trust the keyGranter to do the correct thing here.  That's a trusted role - much more so than a keyManager of course.

e.g. if they have 1 day left on the key and they use Unlock CC to reup another month, we should have the ability to take the keyManager role to prevent chargeback issues.

Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinGrantKeys.sol]: Since the keyGranter sets an expiration instead of adding time to it they already have permissions to revoke keys for all keyOwners without any refund. So this change does not seem to add more risk.

File: smart-contracts/contracts/mixins/MixinKeys.sol (hunk 6 above)
Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: This is minor of course, but I mean this is the same `msg.sender == keyManagerOf[_tokenId] || isLockManager(msg.sender)`

File: smart-contracts/contracts/mixins/MixinGrantKeys.sol (hunk 5 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinGrantKeys.sol]: hmm.  the keyGranter will be able to(as per our current permissions doc) expire any key, so I agree that this doesn't seem to pose any additional risk.

File: smart-contracts/contracts/mixins/MixinKeys.sol (hunk 3 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: I think we had already migrated a handful of getters to public variables a ways back, so maybe go that way for consistency... 
one thing I should check is the size-diff between a public variable and a public getter function. NOte that if we think the benefits to using private vars outweighs the (assumed) size penalty I'm not against going that route.

Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinKeys.sol]: I'll open a new issue - seems drafting a style guide for questions like this would be useful.

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 8 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: the check for this block is:
`if (iDTo == 0) {...`
ensuring this only operates on new keys.

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 10 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: the check for this block is:
`if (toKey.tokenId == 0) {...`
ensuring this only operates on new keys.

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 9 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: Yeah we can simplify this. As we don't have a getter function for keyManager anymore, it probably makes the most sense to say that for all functions:
- if `address(0)` is passed as the keyManager, then the key owner is also the manager.(ie: `grantKeys`)
- if there's no param for keyManager in a function, then the key owner is also the manager.(ie: `shareKey`)

Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: >Thin keyManager does not change in this scenario. And no keyManager means that the keyOwner has those rights.

I just want to take an extra look at this. in this scenario, the existing keyManager is the one sharing the key. would you like to see the keyManager of the "shared" key here be the existing keyManager(of the parent key) or the keyOwner(of the child key)?

**Comment by nfurfaro** [comment]
Closing as this is now implemented.
